Tap Go Library

jcalabro opened the pull request:
This complements #1170 and https://github.com/bluesky-social/atproto/pull/4290 by creating a new go library for working with the `tap` websocket.

This explicitly does not create HTTP handler helpers, but the `tap.Event` type can be used to support the plain HTTP request/response paradigm supported by `tap`.

It's pretty important that the  websocket reconnection and message retry logic be robust, if ever want to say replace inferno with this library, so I went ahead and did what feels like over-engineering, but I don't actually think it is. I would not want to ship this to our prod appview backend without very robust retry logic that mimics what we already do internally (infinite retry on transient processing failures).

There are mocked tests here, and I'm using this library in a new system I'm working on and it appears to be working well (not public yet, but it will be! :eyes:)

bnewbold committed:
lexicon tests: avoid reserved keywords

dholms committed:
fix magic numbers in outbox

dholms committed:
add outbox parallelism cfg & clean up config

dholms committed:
Make repo fetch timeout configurable in `sync-tool` (#1192)

Some big repos take more than 30 seconds to fetch as I have learned the
hard way.

dholms committed:
fix linting error in main

dholms committed:
identity cache size config

dholms committed:
autotrack repos that have a signal collection evt come through

dholms committed:
add type to ws response msgs

dholms committed:
tidy

dholms committed:
fix lint

dholms committed:
add note to readme

dholms committed:
add to build list

bnewbold committed:
new lexgen implementation

bnewbold committed:
new lexgen package (#1195)

This is an effort to get some of the lexgen refactor merged as a
package.

I want to start using this code from `glot`, but am not ready to run and
merge updates to the existing / "legacy" code in `api/`.

dholms committed:
fix deafult time

dholms committed:
fix outbox cfg

bnewbold committed:
lexicon: don't try to validate tokens in data

bnewbold committed:
lexicon: clearer file naming

bnewbold committed:
lexicon: more type field checks

bnewbold committed:
more complete checks for schema types in different places

bnewbold committed:
check that record key type is defined

bnewbold committed:
check for closed empty union

bnewbold committed:
ensure that subscription message schema is a union structurally

bnewbold committed:
improve test coverage

bnewbold committed:
fix lexgen for SchemaMessage change

bnewbold committed:
lexicon: don't allow 'unknown' in params

bnewbold committed:
fix test cases and error msg for no-unknown-params

ThisIsMissEm committed:
Update cmd/sonar/Dockerfile

ThisIsMissEm committed:
Update cmd/supercollider/Dockerfile

ThisIsMissEm committed:
Update cmd/sonar/Dockerfile

dholms committed:
outbox: only set lastId if no error

dholms committed:
start event enumeration from first ID in db rather than 0

dholms committed:
switch between parallel/serial event loading in event cache based on result set size for sparsely populated areas of table

dholms committed:
actually send events in loadEventsSerial

dholms committed:
grabbing some tweaks from whys branch

dholms committed:
remove duplciate sleep

bnewbold committed:
remove breaking resolveLexicon (for now)

bnewbold committed:
sync lexicons from atproto repo

bnewbold committed:
run lexgen to remove deprecated prioritizeFollowedUsers

bnewbold committed:
run lexgen for recent small changes

dholms committed:
fix: store rev on resync record evts

dholms committed:
batch deletes from cache

dholms committed:
max cache size

dholms committed:
update ordering comment

dholms committed:
putting together new event manager

dholms committed:
namign

dholms committed:
switching out event processor impl

bnewbold committed:
better error for account/identity resources; and more test coverage

bnewbold committed:
have mimetype tester handle */*

bnewbold committed:
add description to permission-set; disallow repo col wildcard in perm set

bnewbold committed:
lexicon SDK: stricter schema parsing (#1197)

These are mostly to align with recent changes documented here:
https://github.com/bluesky-social/atproto/discussions/4343

A couple other nits and corner-cases also turned up.

bnewbold committed:
cborgen

bnewbold committed:
lexgen 2025-11-23 (#1200)

This is still with the older lexgen code. Just keeping things
synchronized until new stuff is confidently ready.

estrattonbailey committed:
Add lexicon defs

estrattonbailey committed:
[AAv2] Add lexicon defs (#1201)

New age assurance laws are coming into effect soon, and Bluesky needs to
update its systems to remain legally compliant. Very soon, we'll be
introducing pathways of integration with our upstream provider,
[KWS](https://www.kidswebservices.com/en-US). These will be more
flexible, and allow us to handle a broader range of regions and required
rules.

This PR is the first step: the types and rulesets we'll use for this
next phase. This is the sister PR to
https://github.com/bluesky-social/atproto/pull/4363

dholms committed:
move more over to new event manager

bnewbold committed:
lexicon: remove 'null' def support

bnewbold committed:
no 'blob' permissions in permission-set

bnewbold committed:
lexicon: rpc perm aud can't have DID in permission-set

bnewbold committed:
minor error message improvements

bnewbold committed:
lexgen updates for SchemaNull removal

dholms committed:
wip

dholms committed:
switch outbox over to using event manager

bnewbold committed:
atproto/lexicon: more tweaks from spec updates (#1202)

dholms committed:
backpressure on resyncs

dholms committed:
clean up outbox

bnewbold committed:
bump go to v1.25

bnewbold committed:
update dockerfiles for go v1.25 and debian bookworm

bnewbold committed:
update CI to go v1.25

dholms committed:
repo manager, resync buffer, refactor resyner

dholms committed:
refactor firehose & nexus instantiation"

dholms committed:
clean up event mangaer waiting code

dholms committed:
tweak event cache params

dholms committed:
do better about keeping cache size to rough size

dholms committed:
refactor resync buffer

dholms committed:
satisfy list repos with JOIN

dholms committed:
Satisfy sync.listRepos with JOIN (#1204)

This fixes a bug where `sync.listRepos` returns a 500 if we fail to read
the current commit from the DB.

Also fixes a long-standing TODO around satisfying this method with a
JOIN rather than repeated SELECTs.

dholms committed:
nexus -> tap

bnewbold committed:
lexgen: decrease logging verbosity

bnewbold committed:
lexgen: decrease logging verbosity (#1206)

bnewbold committed:
update go to v1.25 (#1203)

bnewbold committed:
add lexlint package

bnewbold committed:
add lexlint package (#1208)

Use for `goat lex lint ...`.

This is maintained here in indigo so other projects can use it, and so
that it can be maintained as SDK and protocol changes.

bnewbold committed:
relay: migrate from xrpc pkg to atclient

bnewbold committed:
update lexicon resolution to use atclient pkg

bnewbold committed:
update lexicon resolution to use atclient pkg (#1209)

dholms committed:
admin token

bnewbold committed:
relay: Fix initial sequence number being lost on restart after initialization (#1185)

If a relay starts with the last log file being empty, it will reset its
sequence number to 0.

In addition, and relatedly, if a new relay is immediately restarted, it
will ignore the `--initial-seq-number` setting.

Fixes #1133

bnewbold committed:
collectiondir: use base64url encoding for cursor param

bnewbold committed:
relay: log unexpected message type

dholms committed:
add did resolution endpoint

bnewbold committed:
relay: expose current seq as metrics guage

bnewbold committed:
relay: log unexpected message type (#1212)

Closes: https://github.com/bluesky-social/indigo/issues/1083

bnewbold committed:
relay: migrate from xrpc pkg to atclient (#1210)

part of both deprecating the old xrpc package, and cleaning up relay
codebase.

bnewbold committed:
relay: expose current seq as metrics guage (#1213)

bnewbold committed:
relay: enforce initial seq as a minimum

dholms committed:
stats

bnewbold committed:
update HostChecker to use models.AccountStatus types

bnewbold committed:
relay: simplify identity evt logic

Always purge local identity account for every #identity event, and pass
through identity events from any source host (not only the current PDS
host).

Previously we were checking the account's PDS hosting location before
processing these. But if an old PDS emited we'd still want to purge the
cache and have the event go through (purging downstream hosts).

We might need to add rate-limits or other anti-abuse on this at some
point. In particular, the handle in the identity event is now *really*
not trusted.

bnewbold committed:
have EnsureAccountHost try identity update on failure

bnewbold committed:
add EnsureAccountStatus, which re-checks upstream account status if needed

bnewbold committed:
clarifying comment

bnewbold committed:
Add missing container image labels + fix licenses label (#1193)

This was noted in #1158

bnewbold committed:
filter handle in identity events

bnewbold committed:
clarify initialSeq behavior in docs/comments

bnewbold committed:
also check handle against DID document

bnewbold committed:
flexibly parse old non-base64url cursors

bnewbold committed:
collectiondir: use base64url encoding for cursor param (#1211)

This should fix a corner-case bug with query param encoding with Swift.

Will need a deploy.

Closes: https://github.com/bluesky-social/indigo/issues/1084

bnewbold committed:
relay: iterate on verification of account host and status (#1215)

This is intended to fix the stuck account status described in:
https://github.com/bluesky-social/indigo/issues/1161

The previous logic was:

- all events were checked against cached hosting location, and event
dropped if not from expected account PDS
- if an account had upstream inactive status, that status would persist
until a new `#account` event was emitted
- if an event is received from a host, and that host doesn't match the
cached account location (host), code will check if the account location
needs to be updated

The new logic:

- identity events immediate purge the cache for that account, and
generally pass-through for any account from any host. This is to ensure
that identity events always "get through" and purge caches, especially
if they are the first event from "new" PDS during a migration
- if an event is received from a host, and the host doesn't match, the
identity cache will be purged and the host checked again with the newly
resolved identity
- if an event is received from the correct host, and the account has
inactive status upstream, the relay will check for the current account
status on the PDS host to check the upstream status (and will update the
database/cache if needed)

It is still the case that:

- account host information never gets created or updated unless an event
is received from the current active host. that is, an event from the
"old" PDS host during a migration is not sufficient; the "new" PDS host
needs to emit an event (any event will do) and have it received by the
relay. This is partially to prevent accounts from claiming to be hosted
on a PDS which they are not (bi-directional verification)
- an "old" PDS host could emit an `#account` event marking an account as
inactive. If there was no `#identity` event processed before this, this
could result in the account status getting updated. however, the new
logic should make it much more likely for the account status to
"converge", and for that to happen without events getting dropped
- an account can get "stuck" active at the relay. If a PDS marked an
account inactive (eg, takendown), but the relay never received that
event (dropped, bug, or connected later), then it will have active
account status at the relay, and will pass-through events. This is a
pretty corner-case situation: a PDS presumably should not be sending
such events if the account is inactive.

cc: @uniphil

bnewbold committed:
relay: enforce initial seq as a minimum (#1214)

bnewbold committed:
syntax: switch AtIdentifier to be string like all other types

bnewbold committed:
code updates for string-like AtIdentifier

dholms committed:
use admin token on webhook client

dholms committed:
add repo info route

dholms committed:
dockerfile & action

dholms committed:
Merge branch 'main' into sync-tool

dholms committed:
upgrade cli & versioninfo libs

bnewbold committed:
lexicon: subscription msg is required

bnewbold committed:
implement lex tweak

dholms committed:
readme cleanup

dholms committed:
pr feedback

dholms committed:
add back in verifyStructure to sync lib

dholms committed:
run crawler on a loop

dholms committed:
fix crawler run call

dholms committed:
rm jaeger & add prometheus metrics

dholms committed:
admin token -> api token

dholms committed:
logging middleware

bnewbold committed:
syntax: datetime syntax examples from interop tests

bnewbold committed:
syntax: check for negative datetimes

bnewbold committed:
lexicon: validate schema NSID in CheckSchema

bnewbold committed:
update interop test vectors

bnewbold committed:
mst: interop tests for example key height

dholms committed:
remove unused log

bnewbold committed:
update MST interop tests fixtures with correct key heights

rafaeleyng committed:
lexgen getRelationships

rafaeleyng committed:
lexgen getRelationships (#1224)

bnewbold committed:
mst: interop tests for example key height (#1222)

Pulls in fixed MST key height examples, and adds a simple test to
confirm heights.

See: https://github.com/bluesky-social/atproto-interop-tests/pull/11

bnewbold committed:
lexicon: validate schema NSID (#1221)

Thought this was already getting checked!

Caught by @matthieusieben over at:
https://github.com/bluesky-social/atproto-interop-tests/pull/8

bnewbold committed:
syntax: clarify behaviors around year zero (#1220)

See: https://github.com/bluesky-social/atproto-website/pull/479

dholms committed:
rm otel

dholms committed:
readme & config revamp

dholms committed:
use xsync.Map

dholms committed:
atomic.bool

dholms committed:
repo fetch timeout tweak

dholms committed:
locking in event manager

dholms committed:
propagate context in gorm calls

dholms committed:
fix lint errs

dholms committed:
tidy go.mod

dholms committed:
fix docker build - update go verison

rafaeleyng committed:
contact import lexicons

rafaeleyng committed:
contact import lexicons (#1225)

dholms committed:
fix env reading

dholms committed:
api token -> admin password

dholms committed:
user evt -> identity evt

dholms committed:
only create on add repos, dont overwrite

dholms committed:
fmt

dholms committed:
tag Docker based on git tags

dholms committed:
retry timeout config

dholms committed:
fix docker build again

dholms committed:
put the actual running behind `run` in the clid

dholms committed:
change mount volume to /data

dholms committed:
railway deploy guide

dholms committed:
one more link

dholms committed:
Tap (sync tool) (#1170)

Sync tool for backfilling and consuming the firehose in order to hand
off a fully authenticated stream of operations to an application.

More details in the included README.

TS client lib: https://github.com/bluesky-social/atproto/pull/4290/files

gaearon committed:
Add tap to README

Typed on mobile, have not tested link

dholms committed:
Add tap to README (#1226)

Typed on mobile, have not tested link

jcalabro committed:
sketch out types for the go tap library

jcalabro committed:
implement tap websocket connection flow

bnewbold committed:
update interop commit fixtures with correct CID type

bnewbold committed:
update 'raw' CIDs to 'dag-cbor' CIDs

bnewbold committed:
repo: updated MST structural tests (#1223)

This is on top of: https://github.com/bluesky-social/indigo/pull/1222

Tries to also update the MST structural diff tests with corrected
example keys.

These updated fixtures currently work in this Go implementation, but not
in the TS implementation:
https://github.com/bluesky-social/atproto/pull/4421

damz committed:
cmd/tap: Fix typo in README

bnewbold committed:
oauth docs: replace transitional OAuth scope

bnewbold committed:
cmd/tap: Fix typo in README: TAP_FULL_NETWORK_MODE > TAP_FULL_NETWORK (#1230)

boltlessengineer committed:
cmd/tap: add `plc-url` flag to `tap`

Signed-off-by: Seongmin Lee <git@boltless.me>

bnewbold committed:
oauth docs: replace transitional OAuth scope (#1232)

I think this is the last place we use `transition:generic` in doc
examples.

We could consider switching these to `include:app.bsky.authCreatePosts`,
but I think it is fine either way.

bnewbold committed:
update README

bnewbold committed:
update README (#1235)

should probably do a broader update of this README soon, but for now
just update listed commands and packages to point to newer/recommended
versions.

bnewbold committed:
lexicon: subscription message field is required (#1219)

Spec updated in:
https://github.com/bluesky-social/atproto-website/pull/475

dholms committed:
Tap: allow custom plc directory (#1234)

One might need custom plc directory endpoint other than
<https://plc.directory> for testing purposes or for plc mirrors.

As tap supports full network backfill, it would be better to test that
on sandboxed atproto infra than public infra with ~40M repos.

So I added a `plc-url` flag and `TAP_PLC_URL` env var.
Same as relay, `ATP_PLC_HOST` env var can also be used instead.

bnewbold committed:
relay: hint PG for perf on listRepos

As indicated in the comment, adding an actual foreign key relationship
would probably help with this. But doing it as a query tweak is a
simpler change to deploy (no schema change).

jcalabro committed:
Catch Panics in EventManager.broadcastEvent

jcalabro committed:
Catch Panics in EventManager.broadcastEvent (#1240)

Seeing lots of crashes in rainbow due to us sending to `s.outgoing` even
though it's already been closed.

This is clearly a bandaid, but we need a fix ASAP, and this should
probably be in there anyways

<img width="1608" height="823" alt="image"
src="https://github.com/user-attachments/assets/ce0ba2d3-b9a4-4564-9de9-3b69070ef67b"
/>

^ note that large recent missing data blip was me cutting the rainbow
process over to a new management system. Had a bit of downtime (sorry!)

jcalabro committed:
polish up the websocket consumer

jcalabro committed:
add some tests

jcalabro committed:
clean up the tests some

jcalabro committed:
add docs and a json marshaling/unmarshaling test for events

jcalabro committed:
expose ACK publicly

jcalabro committed:
just use a single event type, no need for a dedicated ACK

jcalabro committed:
self-review

jcalabro committed:
rm unused function

jcalabro committed:
better handling of message retries

jcalabro committed:
write a test to ensure that non-retryable errors are not retried indefinitely on error

jcalabro committed:
wrong event type name

jcalabro committed:
fix retry logic

jcalabro committed:
more fixes

jcalabro committed:
fix double increment bug

jcalabro committed:
write a godoc comment for the tap package

jcalabro committed:
namespace all builds

dholms committed:
add user-agent to all tap requests

dholms committed:
verify signature on resync

dholms committed:
Add user-agent to all tap requests (#1245)

dholms committed:
respect backoff in resyncer

dholms committed:
Tap: respect backoff in resyncer (#1247)

Resyncer was setting `retry_after` but then not actually respecting it
when claiming a job

Closes https://github.com/bluesky-social/indigo/issues/1239

dholms committed:
Tap: verify signature on resync (#1246)

We were verifying structure of commit, but not verifying the signature
itself.

Closes https://github.com/bluesky-social/indigo/issues/1238

dholms committed:
retry errored repos

dholms committed:
Tap: retry errored repos (#1248)

Follow up to https://github.com/bluesky-social/indigo/pull/1247

Ensure we're retrying errored repos and increase backoff to 1-60min as
opposed to 1-60s

bnewbold committed:
identity: respect HTTPS_PROXY env var for resolutions

bnewbold committed:
respect HTTPS_PROXY for services

oddgarden6465 committed:
typo in indigo/cmd/sonar/README.md

heathdutton committed:
tap: restore nextID counter on restart to prevent unique constraint errors

Fixes #1253

bnewbold committed:
typo in indigo/cmd/sonar/README.md (#1251)

just a tiny typo fix

cuducos committed:
Adds `recId` to `getSuggestUsers`

cuducos committed:
Adds `recId` to `getSuggestUsers` (#1263)

Reflects the changes from bluesky-social/atproto#4440.

dholms committed:
Tap: restore nextID counter on restart to prevent unique constraint errors (#1257)

When Tap restarts, the EventManager loads existing events from the
database but doesn't initialize the nextID counter. This causes new
events to be assigned IDs starting from 0, which collide with existing
database records.

The fix restores the nextID counter as events are loaded, ensuring new
IDs are always greater than any existing database IDs.

Fixes #1253

dholms reviewed (commented):
Nice this looks good to me! Thanks for putting this together.

Before merging, we should add at least a barebones README. It could just have the example from `doc.go` in it. But more importantly, it should link to `./cmd/tap` which has the fuller documentation for tap. 

I can just see devs looking for tap, coming across this first (since it's in the root dir) and not knowing what to do with it. Actually to that end, maybe we rename this package `tap-client` or similar?

bnewbold commented on tap/websocket.go:
```diff
@@ -0,0 +1,320 @@
+// A thin error wrapper that indicates to the tap client consumer loop that a message
+// should not be retried (i.e. invalid user input that will surely fail again on retry).
+type NonRetryableError struct {
+	err error
```
why not expose the inner error? I feel like there is probably an idiom around this sort of error wrapper which i'm not remembering.

bnewbold reviewed (commented):
thanks for hacking on this!

I think especially for a package like this (vs a daemon), it is good/fine to do just doc.go, which renders nicely on pkg.go.dev. Might be worth a small README which just links to that, I think the current doc.go could be improved though: link to the `tap` command README, or even the announcement blog post. Have a section talking about how to do clean shutdown. Maybe talk about some design properties, like how backpressure is expected to work, what kind of performance (evt/sec) is expected with/without ACKs, etc.

I'm not super in to having the Makefile build the binaries in each cmd directory. I'm open to rethinking how it works, but I don't think the current setup is obviously wrong, and think we should discuss it in a separate PR not here.

I think the module should be named `tapclient`, not `tap`. I'm kind of ambivalent about all the options of where to put it: top level is fine ok but gets busy; `cmd/tap/tapclient` isn't idiomatic or accessible but feels a bit tidier; maybe `pkg/tapclient` but if we do that we should be more consistent. top-level is probably best for now.

I'd be curious to iterate on this more together, but also think this is good and helpful as is and don't think we need to block merging.

rootdiae committed:
update copyright year to 2026

rafaeleyng committed:
Add drafts lexicon

cuducos committed:
Revert "Adds `recId` to `getSuggestUsers`"

This reverts commit 2c00289a95bd0f6b06ae08e0ac7b80fbbc8b367d.

rafaeleyng committed:
Drafts lexicon (#1270)

johnspurlock-skymethod committed:
Add tap executable to .gitignore

cuducos committed:
Revert "Adds `recId` to `getSuggestUsers`" (#1271)

This reverts commit 2c00289a95bd0f6b06ae08e0ac7b80fbbc8b367d.


We need to remove this field here and from `@atproto/bsky`, deploy the
lexicon without it, and then re-add it as string.

Long story short: `recId` is an `int64` and TS/JS cannot get the machine
ID out of it because it can't handle big integers unless with explicity
`BigInt` — which is not the case in our stack. So we're gonna pass it
ias string instead.

Related PRs:

* bluesky-social/atproto#4440
* #1263
* bluesky-social/seeemore#155 
* bluesky-social/seeemore#158

davecheney committed:
pds: remove duplicated import

github.com/bluesky-social/indigo/api/atproto was imported twice, the second time
under an alias. This commit removes the duplicated import.

bnewbold committed:
update copyright year to 2026 (#1269)

bnewbold committed:
pds: remove duplicated import (#1273)

github.com/bluesky-social/indigo/api/atproto was imported twice, the
second time under an alias. This commit removes the duplicated import.

bnewbold committed:
Add tap executable to .gitignore (#1272)

bnewbold committed:
have custom `http.Transport` respect HTTPS_PROXY env var (including identity resolution in SDK) (#1249)

This came up in: https://github.com/bluesky-social/goat/issues/22

For things like `goat`, we definitely want to respect this setting. And
should be consistent about this elsewhere.

I don't want to rush this before the holidays, so it might need to wait
until mid-January.

cc'ing @jcalabro for eyes on this in terms of production impact. I know
some of our environments use HTTP proxying, and this could have an
impact when merged and deployed.

bnewbold committed:
relay: hint PG for perf on listRepos (#1237)

As indicated in the comment, adding an actual foreign key relationship
would probably help with this. But doing it as a query tweak is a
simpler change to deploy (no schema change).

I have tested this in prod and it helped a bunch with perf. I have not
tested even locally with an actual call to listRepos, should do that
before merging+deploying.

closes: https://github.com/bluesky-social/indigo/issues/1236

cuducos committed:
Adds `recID` to suggested user response (as string)

cuducos committed:
Adds `recID` to suggested user response (as string) (#1275)

Re-adds `recId` similar to #1263, but as string (see #1271 for context)

bnewbold committed:
`syntax.AtIdentifier` as simple string (#1218)

This was a clever struct wrapped, but that was weird and inconsistent
with all the other types.

This updates it to be a simple string alias type.

This is (slightly) breaking, and will require updates in downstream
packages. Particularly those doing OAuth.

karitham committed:
atproto/atclient: password auth refresh should use method post

bnewbold committed:
atproto/atclient: password auth refresh should use method post (#1278)

Fix `PasswordAuth#Refresh` using the wrong method.

dholms committed:
Tap: fix repo commit metadata tracking

bnewbold committed:
atclient: allow pw auth to inactive account

bnewbold committed:
atclient: allow pw auth to inactive account (#1281)

This check was overly aggressive. There are a lot of situations where an
account needs to authenticate in "deactivated" state.

I'm not even sure if a log line is necessary here, but figure this is a
change in SDK behavior so it should at least be a bit vocal.

This also fixes a bug where a pointer to the status string was printed,
instead of the status string itself.

davecheney committed:
cmd/tap: use LoadOrCompute when delivering event to worker

The majority of deliverEvent was managing the case where a did worker
was active. As the reference to these workers are already maintained in
an xsync.Map, we can outsource the paperwork of looking up or creating
the worker to Map's LoadOrCompute method.

LoadOrCompute does hold a write lock over the Map during the compute
function, which is a potenial bottleneck, althought the lock time is
O(1) for the allocation of the &DIDWorker so I am hoping for now.

dholms committed:
check prev data when draining the resync buffer

dholms committed:
Tap: fix repo commit metadata tracking (#1280)

Fixes a few issues in Tap. The most important is that we were not
updating prev data & rev when we didn't need to process any ops from a
commit. This would cause commits to appear as disjunctions in commit
history.

This also rearranges some logic in `ProcessCommit` which would cause us
to mark a repo "desynced" if an event came in while resyncing rather
than adding that event to the resync buffer.

For events that we add to the resync buffer, we hold off checking that
they play neatly on the current state of the repo (through prevData)
until we actually go to play them back.

axfelix committed:
add install note to readme

davecheney committed:
.golangci.yaml: warn if go.mod/go.sum is dirty

Settin modules-download-mode: readonly causes the linter to complain,
rather than silently letting the tooling fix things up behind the
scenes. This is a good belt and suspenders check to ensure whats in the
commit is what is built at that commit.

davecheney committed:
workflows: use go.mod as source of truth for the Go version to use

Rather than mentioning the Go version in several places,
actions/setup-go can source the correct version from the go.mod file,
removing one place where they can get out of sync.

I wasn't sure if the go-version-file syntax is available with setup-go
v4, so I went ahead and bumped up the checkout and setup actions to
their latest versions. I use these versions on several projects as they
are what dependabot will upgrade you too by default, so feel pretty
confident this is a safe change.

axfelix committed:
add install note to readme (#1284)

dholms committed:
Tap: fix nil pointer err on prevData

dholms committed:
Tap: fix nil pointer err on prevData (#1287)

Prev data can be nil in the case of legacy (non sync1.1) commits,
causing a nil pointer error.

davecheney committed:
cmd/tap: move ownership of ^C'able context outside runTap

Rather than making runTap responsible for wrapping the context.Context
passsed to it in a signal handler, push that responsibility to the
caller as SIGTERM'ness is really a property of the process wrapped
around tap, rather than runTap itself managing its own lifecycle. The
logic is now runTap is pass a parent context, when that context is done,
everyone shuts down. The why it was cancelled is not something runTap
has to manage.

bnewbold committed:
workflows: use go.mod as source of truth for the Go version to use (#1286)

Rather than mentioning the Go version in several places,
actions/setup-go can source the correct version from the go.mod file,
removing one place where they can get out of sync.

I wasn't sure if the go-version-file syntax is available with setup-go
v4, so I went ahead and bumped up the checkout and setup actions to
their latest versions. I use these versions on several projects as they
are what dependabot will upgrade you too by default, so feel pretty
confident this is a safe change.

jcalabro committed:
.golangci.yaml: warn if go.mod/go.sum is dirty (#1285)

Setting modules-download-mode: readonly causes the linter to complain,
rather than silently letting the tooling fix things up behind the
scenes. This is a good belt and suspenders check to ensure whats in the
commit is what is built at that commit.

jcalabro committed:
cmd/tap: use LoadOrCompute when delivering event to worker (#1283)

The majority of `deliverEvent` was managing the case where a did worker
was active. As the reference to these workers are already maintained in
an xsync.Map, we can outsource the paperwork of looking up or creating
the worker to Map's LoadOrCompute method.

LoadOrCompute does hold a write lock over the Map during the compute
function, which is a potential bottleneck, althought the lock time is
O(1) for the allocation of the &DIDWorker so I am hoping this is ok for
now.

davecheney committed:
cmd/tap: move ownership of ^C'able context outside runTap (#1288)

Rather than making runTap responsible for wrapping the context.Context
passsed to it in a signal handler, push that responsibility to the
caller as SIGTERM'ness is really a property of the process wrapped
around tap, rather than runTap itself managing its own lifecycle. The
logic is now runTap is pass a parent context, when that context is done,
everyone shuts down. The why it was cancelled is not something runTap
has to manage.

jcalabro committed:
merge w/ main
